packages/sdk/client-services/src/packlets/services/client-rpc-server.ts

```diff
@@ -0,0 +1,92 @@
+//
+// Copyright 2023 DXOS.org
+//
+
+import { ClientServices } from '@dxos/client';
+import { Any, ServiceHandler, Stream } from '@dxos/codec-protobuf';
+import { raise } from '@dxos/debug';
+import { parseMethodName, RpcPeer, RpcPeerOptions } from '@dxos/rpc';
+import { MaybePromise } from '@dxos/util';
+
+import { ServiceRegistry } from './service-registry';
+
+export type ClientRpcServerParams = {
+  serviceRegistry: ServiceRegistry<ClientServices>;
+  handleCall?: (
+    method: string,
+    params: Any,
+    handler: (method: string, params: Any) => MaybePromise<Any>
+  ) => Promise<Any>;
+  handleStream?: (
+    method: string,
+    params: Any,
+    handler: (method: string, params: Any) => Stream<Any>
+  ) => MaybePromise<Stream<Any>>;
+} & Omit<RpcPeerOptions, 'callHandler' | 'streamHandler'>;
+
+export class ClientRpcServer {
+  private readonly _serviceRegistry: ServiceRegistry<ClientServices>;
+  private readonly _rpcPeer: RpcPeer;
+  private readonly _handlerCache = new Map<string, ServiceHandler<any>>();
+  private readonly _handleCall: ClientRpcServerParams['handleCall'];
+  private readonly _handleStream: ClientRpcServerParams['handleStream'];
+
+  constructor(params: ClientRpcServerParams) {
+    const { serviceRegistry, handleCall, handleStream, ...rpcOptions } = params;
+    this._handleCall = handleCall;
+    this._handleStream = handleStream;
+
+    this._serviceRegistry = serviceRegistry;
+    this._rpcPeer = new RpcPeer({
+      ...rpcOptions,
+      callHandler: (method, params) => {
+        const [serviceName, methodName] = parseMethodName(method);
+        const service = this._getServiceHandler(serviceName);
```

Comment:
should this be called inside the handler? what if the service isn't available here but then is after the wait inside `this._handleCall`? perhaps worth testing missing service + dynamic service?